### PR TITLE
Improve syntax highlighting

### PR DESCRIPTION
**Pull request by jiribenes**
### Changes
- parse comments precisely
- support splices in (currently all) strings
- support individual characters
- support escapes in strings and characters
- support multi-line strings
- support namespace declarations
- parse capability sets precisely
- add folding markers
- support float literals
- parse typed holes in line with the rest of the file
- add operator support (only visible in some themes)
- `in` is now a conditional keyword (rendered as a keyword only in `var x in r = ...`)
- fix definitions being stuck at `=>` and then not parsing the rest of the arguments
- highlight `at` correctly
- highlight invalid escapes
- highlight `module`s and `import`s correctly
- add support for `region`

### Caveats / open questions
- currently, we only highlight `'\uA0EB'` char literals as opposed to `\uA0EB` detected in the parser, see issue https://github.com/effekt-lang/effekt/issues/521

### Future work
- precise parsing of extern defs
- add support for identifiers that use `?` and `!` (I tried, but I couldn't get it to work)
- add a test suite (low-prio, but it would be nice to know we're somewhat stable)
- use semantic highlighting from the compiler via LSP: https://code.visualstudio.com/api/language-extensions/semantic-highlight-guide, then we could even distinguish an effect type from a normal type, a parameter from a variable, etc. :)
  - demo gifs here: https://github.com/Microsoft/vscode-languageserver-node/pull/367#issuecomment-411043236
- support `box {<capabilities>*} { ... }` syntax

### Screenshots (before [left] vs after [right])

Screenshot 2024-07-20 at 09 19 13
Screenshot 2024-07-20 at 09 19 46
Screenshot 2024-07-20 at 09 21 59
Screenshot 2024-07-20 at 09 23 46
Screenshot 2024-07-20 at 10 52 03
Screenshot 2024-07-20 at 14 42 40
Screenshot 2024-07-20 at 18 15 22
Screenshot 2024-07-21 at 18 14 46
Screenshot 2024-07-22 at 10 46 27

### CODE REVIEW

File: syntaxes/effekt.tmLanguage.json [hunk 1]
```diff
@@ -112,46 +110,68 @@
...
+      }]
+    },
+    "characters": {
+      "name": "string.quoted.single.effekt",
+      "begin": "'",
+      "end": "'",
+      "patterns": [{
+        "name": "constant.character.escape.effekt",
+        "match": "\\\\((u[0-9a-fA-F]{4})|.)"
       }]
```
Comment (jiribenes) [inline syntaxes/effekt.tmLanguage.json:136]: I hereby reject the way character escapes are actually parsed, so the grammar is exactly what you would expect (moreover, we're somewhat lax in comparison with the actual lever in order to allow more characters in a char literal, which is -- I think -- a good thing for highlighting)

File: syntaxes/effekt.tmLanguage.json [hunk 2]
```diff
@@ -93,17 +131,62 @@
...
+    },
+    "characters": {
       "patterns": [{
-        "match": "\\b[0-9]+",
-        "name": "constant.numeric"
+        "match": "'(\\\\[nrt\\\\']|[^'\\\\])'",
+        "name": "string.quoted.single.char.effekt"
+      }, {
+        "match": "\\\\u[0-9A-Fa-f]{4}",
+        "name": "string.quoted.single.char.effekt"
```
Comment (jiribenes) [inline syntaxes/effekt.tmLanguage.json]: I was quite surprised, but `\uFF0F` is really a char literal with no need for single quotes!

File: syntaxes/effekt.tmLanguage.json [hunk 3]
```diff
@@ -102,11 +102,20 @@
     },
     "keywords": {
       "patterns": [{
-        "match": "\\b(module|import|def|fun|val|var|effect|type|match|case|record|extern|include|box|unbox|in|region|new|resource|interface|namespace|module|include|export|as)\\b",
+        "match": "\\b(module|import|def|fun|val|var|effect|type|record|extern|include|box|unbox|in|region|new|resource|interface|namespace|module|include|export|as)\\b",
```
Comment (jiribenes) [inline syntaxes/effekt.tmLanguage.json:105]: I moved `match` and `case` to control flow keywords where they should belong according to the usual TextMate scopes. I hope everybody can survive that they might have a different colour now :^)

**Comment by jiribenes** [comment]
I'll keep better support for externs for later, but here's a sneak peek:
Screenshot 2024-07-20 at 09 30 23

File: syntaxes/effekt.tmLanguage.json (hunk 2 above)
Comment (jiribenes) [reply syntaxes/effekt.tmLanguage.json]: Filed https://github.com/effekt-lang/effekt/issues/521 just to be sure that we don't have to support this :)

File: syntaxes/effekt.tmLanguage.json [hunk 4]
```diff
@@ -1,123 +1,258 @@
...
+        "name": "comment.block.effekt"
+      }]
     },
     "definitions": {
       "patterns": [{
-          "begin": "\\s*(extern)?\\s*(\\{[^\\}]*\\}|[a-zA-Z][a-z-A-Z0-9_$]*)\\s*(def)\\s+([a-zA-Z][a-z-A-Z0-9_$]*)\\b",
-          "end" : "[=\\n]",
-          "beginCaptures": {
+          "begin": "\\s*(extern)\\s+((?:(?:\\{[^\\}]*\\}|[a-zA-Z][a-z-A-Z0-9_$]*)\\s+)?)\\s*(def)\\s+([a-zA-Z][a-z-A-Z0-9_$]*)\\b",
+          "end": "(?:(?=\\n)|(?=\\b(?:(?!=>)[^=]>\\b)|(?:(?!>)\\=)))",
```
Comment (jiribenes) [inline syntaxes/effekt.tmLanguage.json:40]: This is wild, but the gist is: an extern is ended by either:
- a newline `\n`
- an equal sign `=` which isn't a part of `=>`

If anybody can make this regex saner, please, let me know 🙏

File: syntaxes/effekt.tmLanguage.json [hunk 5]
```diff
@@ -1,123 +1,258 @@
...
             "1": { "name": "keyword.declaration.var" },
             "2": { "name": "variable" }
-          }
+          },
+          "end": "(?:(?=\\n)|(?=\\b(?:(?!=>)[^=]>\\b)|(?:(?!>)\\=)))",
+          "patterns": [{
+            "match": "\\s+(in)\\s+([a-zA-Z][a-z-A-Z0-9_$]*)\\b",
+            "captures": {
+              "1": { "name": "keyword.syntax.in.effekt" },
+              "2": { "name": "entity.name.region.effekt" }
```
Comment (jiribenes) [inline syntaxes/effekt.tmLanguage.json:73]: Should we have different scopes for regions and capabilities?
Currently we distinguish `entity.name.region.effekt` (the `r` in `var x in r`) vs `variable.other.capability` (everywhere where a cap set can be)

**Comment by jiribenes** [comment]
I've updated the screenshots with some of the newer features, but I'm running out of things to add, so feel free to review and merge :)

File: syntaxes/effekt.tmLanguage.json [hunk 6]
```diff
@@ -108,40 +114,43 @@
...
+        "name": "keyword.control.effect.effekt"
       }, {
         "match": "\\b(while|match|case|if|else)\\b",
         "name": "keyword.control.flow.effekt"
       }, {
         "match": "\\b(is|and)\\b",
         "name": "keyword.operator.effekt"
+      }, {
+        "match": "\\b(extern)\\b",
+        "name": "storage.modifier.extern.effekt"
```
Comment (jiribenes) [inline syntaxes/effekt.tmLanguage.json:153]: I consulted [the TextMate grammar reference](https://macromates.com/manual/en/language_grammars) and it feels that `extern` is more of a `storage.modifier.*` than a `keyword.*`

File: syntaxes/effekt.tmLanguage.json [hunk 7]
```diff
@@ -1,123 +1,286 @@
...
+        "name": "keyword.control.jump.effekt"
+      }, {
+        "match": "\\b(do|try|with)\\b",
+        "name": "keyword.control.effect.effekt"
+      }, {
+        "match": "\\b(while|match|case|if|else)\\b",
+        "name": "keyword.control.flow.effekt"
+      }, {
+        "match": "\\b(is|and)\\b",
+        "name": "keyword.operator.effekt"
```
Comment (jiribenes) [inline syntaxes/effekt.tmLanguage.json:151]: As you can see, we distinguish between `syntax`, `control`, and `operator` keywords (as per https://macromates.com/manual/en/language_grammars; can have different highlighting like in the Dark+ theme in VSCode)

However, which keywords should be in which group?
- I don't know about `as` (since we don't use it, I think?), but it might go into `keyword.operator` if it's meant for pattern matching?
- Where does `resume` go? `keyword.control.jump` or `keyword.control.effect`?
- Where do `in` and `region` go? Should they stay as `syntax` or should they move into `control`? (as creating a new region and allocating into it is control-flow specific?)
- Where do `match` and `case` go? I moved them to `control` since it feels like they should be with `if` and `else`, though beware, I can notice the change in the editor now :)

File: syntaxes/effekt.tmLanguage.json (hunk 5 above)
Comment (b-studios) [reply syntaxes/effekt.tmLanguage.json:74]: is the difference one element in the set vs the set?

Comment (jiribenes) [reply syntaxes/effekt.tmLanguage.json:74]: Here the difference is that we highlight the `r` in `var x in r` and `region r` as `entity.name.region`
vs the capabilities in `extern io def` and `Foo at {io, global, baz}` get highlighted as `variable.other.capability`

So in a situation like
```
def prog() = region r {
  var q: Queue[() => Unit at {r, prog}] in r = <>
}
```
the `r` in the capability set is (potentially) highlighted differently than the other mentions of `r`.

Comment (jiribenes) [reply syntaxes/effekt.tmLanguage.json]: I thought about this for a bit and made the following change: https://github.com/effekt-lang/effekt-vscode/pull/18/commits/f6188ecbfb92ecbf79cbae7614cdf728da83fa4d

TLDR: all such capability variables are primarily scope `variable.other.capability.effekt`, if we know for sure that they are regions, we also apply scope `entity.name.region.effekt` so that somebody potentially can make their theme a little better for this very use-case.

File: syntaxes/effekt.tmLanguage.json (hunk 7 above)
Comment (jiribenes) [reply syntaxes/effekt.tmLanguage.json]: For another reference, here's the source of the VSCode Dark+ theme: https://github.com/microsoft/vscode/blob/main/extensions/theme-defaults/themes/dark_plus.json

You can see that [`keyword.control`](https://github.com/microsoft/vscode/blob/cba3b821977fbd1d45eacf0add96cbf9e3865c1e/extensions/theme-defaults/themes/dark_plus.json#L74) is highlighted differently than just [`keyword`](https://github.com/microsoft/vscode/blob/cba3b821977fbd1d45eacf0add96cbf9e3865c1e/extensions/theme-defaults/themes/dark_vs.json#L323)

Comment (jiribenes) [reply syntaxes/effekt.tmLanguage.json]: There's also the question of what to do with `and` and `is`:
Take a look at the `and` here (left = before, right = after) in two different colour schemes:
Screenshot 2024-07-22 at 11 44 53
Screenshot 2024-07-22 at 11 45 08

In the latter scheme (Dark+) which is used a lot by people, the `and` isn't highlighted at all, which seems quite unfortunate.

Comment (jiribenes) [reply syntaxes/effekt.tmLanguage.json]: As agreed on the Effekt group meeting:
- removed `as`
- `resume` is moved to the `.effect` category
- `in` and `region` stay as non-control keywords
- `match` and `case` stay as control keywords
- `import` stays as a non-control keyword
- `and` and `is` are moved to control keywords
